libraries/libopaemem/hash_map.c

```diff
@@ -0,0 +1,291 @@
+// POSSIBILITY OF SUCH DAMAGE.
+#ifdef HAVE_CONFIG_H
+#include <config.h>
+#endif // HAVE_CONFIG_H
+
+#include <opae/log.h>
+#include <opae/hash_map.h>
+#include "mock/opae_std.h"
+
+#ifndef UNUSED_PARAM
+#define UNUSED_PARAM(x) ((void)(x))
+#endif // UNUSED_PARAM
+
+#define __SHORT_FILE__                                    \
+({                                                        \
+	const char *file = __FILE__;                      \
+	const char *p = file;                             \
+	while (*p)                                        \
+		++p;                                      \
+	while ((p > file) && ('/' != *p) && ('\\' != *p)) \
+		--p;                                      \
+	if (p > file)                                     \
+		++p;                                      \
+	p;                                                \
+})
+
+#define ERR(format, ...)                               \
+fprintf(stderr, "%s:%u:%s() **ERROR** [%s] : " format, \
+	__SHORT_FILE__, __LINE__, __func__, strerror(errno), ##__VA_ARGS__)
+
+fpga_result opae_hash_map_init(opae_hash_map *hm,
+			       uint32_t num_buckets,
+			       uint32_t hash_seed,
+			       uint32_t (*key_hash)(uint32_t num_buckets,
+						    uint32_t hash_seed,
+						    void *key),
+			       int (*key_compare)(void *keya, void *keyb),
+			       void (*key_cleanup)(void *key, void *context),
+			       void (*value_cleanup)(void *value, void *context))
+{
+	if (!hm || !key_hash || !key_compare) {
+		ERR("NULL pointer(s)");
+		return FPGA_INVALID_PARAM;
+	}
+
+	memset(hm, 0, sizeof(*hm));
+
+	hm->buckets = (opae_hash_map_item **)
+			opae_calloc(num_buckets,
+				    sizeof(opae_hash_map_item *));
+	if (!hm->buckets) {
+		ERR("calloc() failed");
+		return FPGA_NO_MEMORY;
+	}
+
+	hm->num_buckets = num_buckets;
+	hm->hash_seed = hash_seed;
+	hm->key_hash = key_hash;
+	hm->key_compare = key_compare;
+	hm->key_cleanup = key_cleanup;
+	hm->value_cleanup = value_cleanup;
+
+	return FPGA_OK;
+}
+
+STATIC opae_hash_map_item *
+opae_hash_map_alloc_item(void *key, void *value)
+{
+	opae_hash_map_item *item;
+	item = (opae_hash_map_item *)
+		opae_malloc(sizeof(opae_hash_map_item));
+	if (item) {
+		item->key = key;
+		item->value = value;
+		item->next = NULL;
+	}
+	return item;
+}
+
+fpga_result opae_hash_map_add(opae_hash_map *hm,
+			      void *key,
+			      void *value)
+{
+	uint32_t key_hash;
+	opae_hash_map_item *item;
+	opae_hash_map_item *list;
+
+	if (!hm) {
+		ERR("NULL pointer");
+		return FPGA_INVALID_PARAM;
+	}
+
+	item = opae_hash_map_alloc_item(key, value);
+	if (!item) {
+		ERR("malloc() failed");
+		return FPGA_NO_MEMORY;
+	}
+
+	key_hash = hm->key_hash(hm->num_buckets,
+				hm->hash_seed,
+				key);
+
+	list = hm->buckets[key_hash];
+	if (!list) {
+		// Bucket was empty. Fill it.
+		hm->buckets[key_hash] = item;
+		return FPGA_OK;
+	}
+
+	// Bucket not empty. Do we have a key collision?
+	while (list) {
```

Comment:
Is this step really necessary?

Reply:
In this generalized approach, add means add or replace.  I'd prefer to keep it this way.

Reply:
You could make it a parameter in the initializer to avoid an O(n) search at every insert if the parent can assert that all keys are unique.

Reply:
Added flags parameter to opae_hash_map_init() and flag OPAE_HASH_MAP_UNIQUE_KEYSPACE so that the list walk can be avoided.